src/ngraph/type/bfloat16.hpp

```diff
@@ -14,27 +14,42 @@
 // limitations under the License.
 //*****************************************************************************
 
-//================================================================================================
-// bfloat16 type
-//================================================================================================
-
 #pragma once
 
+#include <cmath>
 #include <iostream>
 #include <memory>
 #include <string>
 #include <vector>
 
+#define ROUND_MODE_TO_NEAREST_EVEN
+
 namespace ngraph
 {
     class bfloat16
     {
     public:
-        bfloat16() {}
-        bfloat16(float value, bool rounding = false);
-        bfloat16(const bfloat16&) = default;
-        bfloat16& operator=(const bfloat16&) = default;
-        virtual ~bfloat16() {}
+        constexpr bfloat16()
+            : m_value{0}
+        {
+        }
+        constexpr bfloat16(float value)
+            : m_value
+        {
```

Comment:
All of the rounding routines are exposed as static functions in the class and there are unit tests which use each of the rounding methods. All methods are tested. I suppose I could delete the ones we aren't using but I am hesitant to do so. They will continue to be tested so they should not rot.

Reply:
I'm very sorry, but I think you should remove all of your `constexpr` attempts.
I totally get what you were trying for, but I don't believe it's working.
In playing around with your code & https://godbolt.org/ I found it wasn't actually being invoked.
constexpr method must be invoked as constexpr upon invocation.
As in:
```
constexpr bfloat16 test(8.0f);
```
From my brief testing I don't believe this will work for you. (Feel free to disagree & disprove).

I found without specifying the statement as `constexpr` it was still generating a bunch of code (when I tried working `constexpr` routines they did not).
Moreover attempting `constexpr` failed. Looks like due to this:
https://akrzemi1.wordpress.com/2012/12/13/constexpr-unions/
https://en.cppreference.com/w/cpp/language/constant_expression
Which basically says that the union trick isn't allowed in a `constexpr` (in that in a `constexpr` you're only allowed to use the _active_ member of the constructed union).

Also of note, when not used as constexpr, the syntactically nicer versions of the union that you've defined (in which you provide nice constructors) causes more code to be generated than the uglier (but faster) approach of declaring the constructing member as the first element of the union (as in my `UIntToFloatUnion` example above).

As it stands doing a proper `constexpr` bfloat construction looks to be uglier than I have stomach for. Witness this discussion & blog post (not specific to bfloat but you can see the pain):
https://news.ycombinator.com/item?id=11370863
http://brnz.org/hbr/?p=1518

Reply:
I removed the use of unions and simply cast